sdk/python/kfp/containers/entrypoint.py

```diff
@@ -0,0 +1,321 @@
+      raise ValueError('Either value or both metadata_file and field_name '
+                       'needs to be provided. Get value={value}, field_name='
+                       '{field_name}, metadata_file={metadata_file}'.format(
+          value=value,
+          field_name=field_name,
+          metadata_file=metadata_file
+      ))
+    if value is not None:
+      self._value = value
+    else:
+      # Parse the value by inspecting the producer's metadata JSON file.
+      self._value = entrypoint_utils.get_parameter_from_output(
+          metadata_file, field_name)
+
+    self._metadata_file = metadata_file
+    self._field_name = field_name
+
+  # Following properties are read-only
+  @property
+  def value(self) -> Union[float, str, int]:
+    return self._value
+
+  @property
+  def metadata_file(self) -> str:
+    return self._metadata_file
+
+  @property
+  def field_name(self) -> str:
+    return self._field_name
+
+
+class InputArtifact(object):
+  """POD that holds an input artifact."""
+
+  def __init__(self,
+      uri: Optional[str] = None,
+      metadata_file: Optional[str] = None,
+      output_name: Optional[str] = None
+  ):
+    """Instantiates an InputParam object.
+
+    Args:
+      uri: The uri holds the input artifact.
+      metadata_file: The location of the metadata JSON file output by the
+        producer step.
+      output_name: The output name of the artifact in producer step.
+
+    Raises:
+      ValueError: when neither of the following is true:
+        1) uri is provided, and metadata_file and output_name are not; or
+        2) both metadata_file and output_name are provided, and uri is not.
+    """
+    if not ((uri and not (metadata_file or output_name) or (
+        metadata_file and output_name and not uri))):
+      raise ValueError('Either uri or both metadata_file and output_name '
+                       'needs to be provided. Get uri={uri}, output_name='
+                       '{output_name}, metadata_file={metadata_file}'.format(
+          uri=uri,
+          output_name=output_name,
+          metadata_file=metadata_file
+      ))
+
+    self._uri = uri
+    self._metadata_file = metadata_file
+    self._output_name = output_name
+
+  # Following properties are read-only.
+  @property
+  def uri(self) -> str:
+    return self._uri
+
+  @property
+  def metadata_file(self) -> str:
+    return self._metadata_file
+
+  @property
+  def output_name(self) -> str:
+    return self._output_name
+
+  def get_artifact(self) -> artifact.Artifact:
+    """Gets an artifact object by parsing metadata or creating one from uri."""
+    if self.metadata_file and self.output_name:
+      return entrypoint_utils.get_artifact_from_output(
+          self.metadata_file, self.output_name)
+    else:
+      # Provide an empty schema when returning a raw Artifact.
+      result = artifact.Artifact(
+          instance_schema=artifact.DEFAULT_ARTIFACT_SCHEMA)
```

Comment:
I think this problem boils down to which one is the source of truth of the schema: 1) compile time declaration, or 2) runtime output from producer.

I think based on our recent discussion they won't make any difference eventually because we'll pursue a schemaless paradigm, but I'm happy to discuss further.